File: apps/mobile/android/app/src/main/java/com/tchat/components/TchatBreadcrumb.kt
package com.tchat.components

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.platform.LocalHapticFeedback
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.tchat.designsystem.Colors
import com.tchat.designsystem.Spacing

/**
 * Breadcrumb navigation component following Tchat design system
 */
@Composable
fun TchatBreadcrumb(
    items: List<TchatBreadcrumbItem>,
    modifier: Modifier = Modifier,
    style: TchatBreadcrumbStyle = TchatBreadcrumbStyle.Default,
    size: TchatBreadcrumbSize = TchatBreadcrumbSize.Medium,
    separator: ImageVector = Icons.Default.ChevronRight,
    maxItems: Int? = null,
    showHome: Boolean = false,
    homeIcon: ImageVector = Icons.Default.Home,
    onItemClick: ((TchatBreadcrumbItem) -> Unit)? = null
) {
    val hapticFeedback = LocalHapticFeedback.current

    val displayItems = remember(items, maxItems) {
        when {
            maxItems == null || items.size <= maxItems -> items
            maxItems <= 2 -> items.takeLast(maxItems)
            else -> {
                val firstItem = items.first()
                val lastItems = items.takeLast(maxItems - 2)
                val ellipsisItem = TchatBreadcrumbItem(
                    id = "ellipsis",
                    title = "...",
                    isClickable = false
                )
                listOf(firstItem, ellipsisItem) + lastItems
            }
        }
    }

    Row(
        modifier = modifier
            .horizontalScroll(rememberScrollState())
            .padding(horizontal = Spacing.sm),
        horizontalArrangement = Arrangement.spacedBy(size.spacing),
        verticalAlignment = Alignment.CenterVertically
    ) {
        // Home icon (if enabled)
        if (showHome) {
            HomeButton(
                icon = homeIcon,
                size = size,
                onClick = {
                    val homeItem = TchatBreadcrumbItem(
                        id = "home",
                        title = "Home",
                        icon = homeIcon,
                        isClickable = true
                    )
                    onItemClick?.invoke(homeItem)
                    hapticFeedback.performHapticFeedback(
                        androidx.compose.ui.hapticfeedback.HapticFeedbackType.LongPress
                    )
                }
            )

            if (items.isNotEmpty()) {
                SeparatorIcon(
                    icon = separator,
                    size = size
                )
            }
        }

        // Breadcrumb items
        displayItems.forEachIndexed { index, item ->
            BreadcrumbItemView(
                item = item,
                isLast = index == displayItems.lastIndex,
                style = style,
                size = size,
                onClick = {
                    item.action?.invoke()
                    onItemClick?.invoke(item)
                    hapticFeedback.performHapticFeedback(
                        androidx.compose.ui.hapticfeedback.HapticFeedbackType.LongPress
                    )
                }
            )

            if (index < displayItems.lastIndex) {
                SeparatorIcon(
                    icon = separator,
                    size = size
                )
            }
        }
    }
}

@Composable
private fun HomeButton(
    icon: ImageVector,
    size: TchatBreadcrumbSize,
    onClick: () -> Unit
) {
    Icon(
        imageVector = icon,
        contentDescription = "Home",
        tint = Colors.textSecondary,
        modifier = Modifier
            .size(size.iconSize)
            .clickable { onClick() }
    )
}

@Composable
private fun BreadcrumbItemView(
    item: TchatBreadcrumbItem,
    isLast: Boolean,
    style: TchatBreadcrumbStyle,
    size: TchatBreadcrumbSize,
    onClick: () -> Unit
) {
    val itemColor = when (style) {
        TchatBreadcrumbStyle.Default -> {
            when {
                isLast -> Colors.textPrimary
                item.isClickable -> Colors.primary
                else -> Colors.textSecondary
            }
        }
        TchatBreadcrumbStyle.Compact, TchatBreadcrumbStyle.Minimal -> {
            when {
                isLast -> Colors.textPrimary
                item.isClickable -> Colors.textSecondary
                else -> Colors.textTertiary
            }
        }
    }

    val backgroundColor = when {
        style == TchatBreadcrumbStyle.Default && isLast -> Colors.surface
        else -> Color.Transparent
    }

    val clickableModifier = if (item.isClickable && !isLast) {
        Modifier.clickable { onClick() }
    } else {
        Modifier
    }

    Row(
        modifier = Modifier
            .then(clickableModifier)
            .background(
                color = backgroundColor,
                shape = RoundedCornerShape(4.dp)
            )
            .padding(
                horizontal = if (style == TchatBreadcrumbStyle.Default) size.spacing else 0.dp,
                vertical = if (style == TchatBreadcrumbStyle.Default) (size.spacing / 2) else 0.dp
            ),
        horizontalArrangement = Arrangement.spacedBy(size.spacing / 2),
        verticalAlignment = Alignment.CenterVertically
    ) {
        // Icon
        item.icon?.let { icon ->
            Icon(
                imageVector = icon,
                contentDescription = null,
                tint = itemColor,
                modifier = Modifier.size(size.iconSize)
            )
        }

        // Title
        Text(
            text = item.title,
            fontSize = size.fontSize,
            fontWeight = if (isLast) FontWeight.Medium else FontWeight.Normal,
            color = itemColor,
            maxLines = 1
        )
    }
}

@Composable
private fun SeparatorIcon(
    icon: ImageVector,
    size: TchatBreadcrumbSize
) {
    Icon(
        imageVector = icon,
        contentDescription = null,
        tint = Colors.textTertiary,
        modifier = Modifier.size(size.iconSize - 2.dp)
    )
}

/**
 * Breadcrumb item data class
 */
data class TchatBreadcrumbItem(
    val id: String,
    val title: String,
    val icon: ImageVector? = null,
    val isClickable: Boolean = true,
    val action: (() -> Unit)? = null
)

/**
 * Breadcrumb style definitions
 */
enum class TchatBreadcrumbStyle {
    Default,
    Compact,
    Minimal
}

/**
 * Breadcrumb size definitions
 */
enum class TchatBreadcrumbSize(
    val fontSize: androidx.compose.ui.unit.TextUnit,
    val iconSize: androidx.compose.ui.unit.Dp,
    val spacing: androidx.compose.ui.unit.Dp
) {
    Small(
        fontSize = 12.sp,
        iconSize = 12.dp,
        spacing = 4.dp
    ),
    Medium(
        fontSize = 14.sp,
        iconSize = 14.dp,
        spacing = 6.dp
    ),
    Large(
        fontSize = 16.sp,
        iconSize = 16.dp,
        spacing = 8.dp
    )
}

/**
 * Utility functions for creating breadcrumbs
 */
object TchatBreadcrumbUtils {
    fun fromPath(
        path: String,
        separator: String = "/",
        onPathClick: ((String) -> Unit)? = null
    ): List<TchatBreadcrumbItem> {
        val components = path.split(separator).filter { it.isNotEmpty() }

        return components.mapIndexed { index, component ->
            val fullPath = components.take(index + 1).joinToString(separator)

            TchatBreadcrumbItem(
                id = fullPath,
                title = component,
                isClickable = true,
                action = { onPathClick?.invoke(fullPath) }
            )
        }
    }

    fun fromNavigation(
        pathComponents: List<String>,
        onNavigateBack: ((Int) -> Unit)? = null
    ): List<TchatBreadcrumbItem> {
        return pathComponents.mapIndexed { index, component ->
            TchatBreadcrumbItem(
                id = index.toString(),
                title = component,
                isClickable = index < pathComponents.lastIndex,
                action = { onNavigateBack?.invoke(index) }
            )
        }
    }
}

// Preview
@Preview(showBackground = true)
@Composable
fun TchatBreadcrumbPreview() {
    Column(
        modifier = Modifier.padding(Spacing.md),
        verticalArrangement = Arrangement.spacedBy(Spacing.lg)
    ) {
        // Default breadcrumb
        TchatBreadcrumb(
            items = listOf(
                TchatBreadcrumbItem(
                    id = "dashboard",
                    title = "Dashboard",
                    icon = Icons.Default.Dashboard
                ),
                TchatBreadcrumbItem(
                    id = "projects",
                    title = "Projects",
                    icon = Icons.Default.Folder
                ),
                TchatBreadcrumbItem(
                    id = "mobile-app",
                    title = "Mobile App",
                    isClickable = false
                )
            ),
            style = TchatBreadcrumbStyle.Default,
            showHome = true
        )

        Divider()

        // Compact breadcrumb
        TchatBreadcrumb(
            items = listOf(
                TchatBreadcrumbItem(
                    id = "users",
                    title = "Users"
                ),
                TchatBreadcrumbItem(
                    id = "profile",
                    title = "Profile"
                ),
                TchatBreadcrumbItem(
                    id = "settings",
                    title = "Settings",
                    isClickable = false
                )
            ),
            style = TchatBreadcrumbStyle.Compact,
            size = TchatBreadcrumbSize.Small
        )

        Divider()

        // Minimal breadcrumb with truncation
        TchatBreadcrumb(
            items = listOf(
                TchatBreadcrumbItem(
                    id = "level1",
                    title = "Level 1"
                ),
                TchatBreadcrumbItem(
                    id = "level2",
                    title = "Level 2"
                ),
                TchatBreadcrumbItem(
                    id = "level3",
                    title = "Level 3"
                ),
                TchatBreadcrumbItem(
                    id = "level4",
                    title = "Level 4"
                ),
                TchatBreadcrumbItem(
                    id = "level5",
                    title = "Current Page",
                    isClickable = false
                )
            ),
            style = TchatBreadcrumbStyle.Minimal,
            maxItems = 4
        )

        Divider()

        // Path-based breadcrumb
        TchatBreadcrumb(
            items = TchatBreadcrumbUtils.fromPath(
                "/projects/mobile-app/src/components"
            ),
            style = TchatBreadcrumbStyle.Default,
            size = TchatBreadcrumbSize.Medium
        )

        Divider()

        // Large breadcrumb with custom separator
        TchatBreadcrumb(
            items = listOf(
                TchatBreadcrumbItem(
                    id = "docs",
                    title = "Documentation",
                    icon = Icons.Default.Description
                ),
                TchatBreadcrumbItem(
                    id = "api",
                    title = "API Reference",
                    icon = Icons.Default.Link
                ),
                TchatBreadcrumbItem(
                    id = "endpoints",
                    title = "Endpoints",
                    isClickable = false
                )
            ),
            style = TchatBreadcrumbStyle.Default,
            size = TchatBreadcrumbSize.Large,
            separator = Icons.Default.ArrowForward,
            showHome = true,
            homeIcon = Icons.Default.Home
        )
    }
}

File: apps/mobile/ios/Sources/Components/TchatBreadcrumb.swift
//
//  TchatBreadcrumb.swift
//  TchatApp
//
//  Created by Claude on 22/09/2024.
//

import SwiftUI

/// Breadcrumb navigation component following Tchat design system
public struct TchatBreadcrumb: View {

    // MARK: - Breadcrumb Types
    public enum BreadcrumbStyle {
        case standard
        case compact
        case minimal
    }

    public enum BreadcrumbSize {
        case small
        case medium
        case large
    }

    // MARK: - Breadcrumb Item
    public struct BreadcrumbItem {
        let id: String
        let title: String
        let icon: String?
        let isClickable: Bool
        let action: (() -> Void)?

        public init(
            id: String,
            title: String,
            icon: String? = nil,
            isClickable: Bool = true,
            action: (() -> Void)? = nil
        ) {
            self.id = id
            self.title = title
            self.icon = icon
            self.isClickable = isClickable
            self.action = action
        }
    }

    // MARK: - Properties
    let items: [BreadcrumbItem]
    let style: BreadcrumbStyle
    let size: BreadcrumbSize
    let separator: String
    let maxItems: Int?
    let showHome: Bool
    let homeIcon: String
    let onItemTap: ((BreadcrumbItem) -> Void)?

    // MARK: - Private Properties
    private let colors = Colors()

    // MARK: - Computed Properties
    private var fontSize: CGFloat {
        switch size {
        case .small: return 12
        case .medium: return 14
        case .large: return 16
        }
    }

    private var iconSize: CGFloat {
        switch size {
        case .small: return 12
        case .medium: return 14
        case .large: return 16
        }
    }

    private var spacing: CGFloat {
        switch size {
        case .small: return 4
        case .medium: return 6
        case .large: return 8
        }
    }

    private var displayItems: [BreadcrumbItem] {
        guard let maxItems = maxItems, items.count > maxItems else {
            return items
        }

        if maxItems <= 2 {
            return Array(items.suffix(maxItems))
        }

        // Show first item, ellipsis, and last items
        let firstItem = items.first!
        let lastItems = Array(items.suffix(maxItems - 2))

        var result = [firstItem]

        // Add ellipsis item
        let ellipsisItem = BreadcrumbItem(
            id: "ellipsis",
            title: "...",
            isClickable: false
        )
        result.append(ellipsisItem)

        result.append(contentsOf: lastItems)
        return result
    }

    // MARK: - Initializer
    public init(
        items: [BreadcrumbItem],
        style: BreadcrumbStyle = .standard,
        size: BreadcrumbSize = .medium,
        separator: String = "chevron.right",
        maxItems: Int? = nil,
        showHome: Bool = false,
        homeIcon: String = "house",
        onItemTap: ((BreadcrumbItem) -> Void)? = nil
    ) {
        self.items = items
        self.style = style
        self.size = size
        self.separator = separator
        self.maxItems = maxItems
        self.showHome = showHome
        self.homeIcon = homeIcon
        self.onItemTap = onItemTap
    }

    // MARK: - Body
    public var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: spacing) {
                // Home icon (if enabled)
                if showHome {
                    homeButton

                    if !items.isEmpty {
                        separatorView
                    }
                }

                // Breadcrumb items
                ForEach(Array(displayItems.enumerated()), id: \.element.id) { index, item in
                    breadcrumbItemView(item, isLast: index == displayItems.count - 1)

                    if index < displayItems.count - 1 {
                        separatorView
                    }
                }
            }
            .padding(.horizontal, Spacing.sm)
        }
    }

    // MARK: - Home Button
    @ViewBuilder
    private var homeButton: some View {
        Button(action: {
            // Handle home tap
            let homeItem = BreadcrumbItem(
                id: "home",
                title: "Home",
                icon: homeIcon,
                isClickable: true
            )
            onItemTap?(homeItem)

            // Haptic feedback
            let impactFeedback = UIImpactFeedbackGenerator(style: .light)
            impactFeedback.impactOccurred()
        }) {
            Image(systemName: homeIcon)
                .font(.system(size: iconSize))
                .foregroundColor(colors.textSecondary)
        }
    }

    // MARK: - Breadcrumb Item View
    @ViewBuilder
    private func breadcrumbItemView(_ item: BreadcrumbItem, isLast: Bool) -> some View {
        Group {
            if item.isClickable && !isLast {
                Button(action: {
                    item.action?()
                    onItemTap?(item)

                    // Haptic feedback
                    let impactFeedback = UIImpactFeedbackGenerator(style: .light)
                    impactFeedback.impactOccurred()
                }) {
                    itemContent(item, isLast: isLast)
                }
                .buttonStyle(PlainButtonStyle())
            } else {
                itemContent(item, isLast: isLast)
            }
        }
    }

    // MARK: - Item Content
    @ViewBuilder
    private func itemContent(_ item: BreadcrumbItem, isLast: Bool) -> some View {
        HStack(spacing: spacing / 2) {
            // Icon
            if let icon = item.icon {
                Image(systemName: icon)
                    .font(.system(size: iconSize))
                    .foregroundColor(itemColor(item, isLast: isLast))
            }

            // Title
            Text(item.title)
                .font(.system(size: fontSize, weight: isLast ? .medium : .regular))
                .foregroundColor(itemColor(item, isLast: isLast))
                .lineLimit(1)
        }
        .padding(.horizontal, style == .standard ? spacing : 0)
        .padding(.vertical, style == .standard ? spacing / 2 : 0)
        .background(
            itemBackground(item, isLast: isLast)
        )
        .cornerRadius(style == .standard ? 4 : 0)
    }

    // MARK: - Separator View
    @ViewBuilder
    private var separatorView: some View {
        Image(systemName: separator)
            .font(.system(size: iconSize - 2))
            .foregroundColor(colors.textTertiary)
    }

    // MARK: - Styling Methods
    private func itemColor(_ item: BreadcrumbItem, isLast: Bool) -> Color {
        switch style {
        case .standard:
            if isLast {
                return colors.textPrimary
            } else if item.isClickable {
                return colors.primary
            } else {
                return colors.textSecondary
            }
        case .compact, .minimal:
            if isLast {
                return colors.textPrimary
            } else if item.isClickable {
                return colors.textSecondary
            } else {
                return colors.textTertiary
            }
        }
    }

    private func itemBackground(_ item: BreadcrumbItem, isLast: Bool) -> Color {
        guard style == .standard else { return Color.clear }

        if isLast {
            return colors.surface
        } else if item.isClickable {
            return Color.clear
        } else {
            return Color.clear
        }
    }
}

// MARK: - Convenience Initializers
extension TchatBreadcrumb {
    public static func fromPath(
        _ path: String,
        separator: String = "/",
        style: BreadcrumbStyle = .standard,
        size: BreadcrumbSize = .medium,
        onPathTap: ((String) -> Void)? = nil
    ) -> TchatBreadcrumb {
        let components = path.split(separator: Character(separator)).map(String.init)

        let items = components.enumerated().map { index, component in
            let path = components.prefix(index + 1).joined(separator: separator)

            return BreadcrumbItem(
                id: path,
                title: component,
                isClickable: true
            ) {
                onPathTap?(path)
            }
        }

        return TchatBreadcrumb(
            items: items,
            style: style,
            size: size
        )
    }
}

// MARK: - NavigationPath Support
@available(iOS 16.0, *)
extension TchatBreadcrumb {
    public init(
        navigationPath: Binding<NavigationPath>,
        pathComponents: [String],
        style: BreadcrumbStyle = .standard,
        size: BreadcrumbSize = .medium,
        separator: String = "chevron.right",
        maxItems: Int? = nil,
        showHome: Bool = true,
        homeIcon: String = "house"
    ) {
        let items = pathComponents.enumerated().map { index, component in
            BreadcrumbItem(
                id: "\(index)",
                title: component,
                isClickable: index < pathComponents.count - 1
            ) {
                // Navigate back to this level
                var newPath = NavigationPath()
                for i in 0...index {
                    newPath.append(pathComponents[i])
                }
                navigationPath.wrappedValue = newPath
            }
        }

        self.init(
            items: items,
            style: style,
            size: size,
            separator: separator,
            maxItems: maxItems,
            showHome: showHome,
            homeIcon: homeIcon
        ) { item in
            // Handle navigation
            if item.id == "home" {
                navigationPath.wrappedValue = NavigationPath()
            }
        }
    }
}

// MARK: - Preview
#if DEBUG
struct TchatBreadcrumb_Previews: PreviewProvider {
    static var previews: some View {
        ScrollView {
            VStack(spacing: Spacing.lg) {
                // Default breadcrumb
                TchatBreadcrumb(
                    items: [
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "dashboard",
                            title: "Dashboard",
                            icon: "chart.bar"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "projects",
                            title: "Projects",
                            icon: "folder"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "mobile-app",
                            title: "Mobile App",
                            isClickable: false
                        )
                    ],
                    style: .standard,
                    showHome: true
                )

                Divider()

                // Compact breadcrumb
                TchatBreadcrumb(
                    items: [
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "users",
                            title: "Users"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "profile",
                            title: "Profile"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "settings",
                            title: "Settings",
                            isClickable: false
                        )
                    ],
                    style: .compact,
                    size: .small
                )

                Divider()

                // Minimal breadcrumb with truncation
                TchatBreadcrumb(
                    items: [
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "level1",
                            title: "Level 1"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "level2",
                            title: "Level 2"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "level3",
                            title: "Level 3"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "level4",
                            title: "Level 4"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "level5",
                            title: "Current Page",
                            isClickable: false
                        )
                    ],
                    style: .minimal,
                    maxItems: 4
                )

                Divider()

                // Path-based breadcrumb
                TchatBreadcrumb.fromPath(
                    "/projects/mobile-app/src/components",
                    style: .standard,
                    size: .medium
                )

                Divider()

                // Large breadcrumb with custom separator
                TchatBreadcrumb(
                    items: [
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "docs",
                            title: "Documentation",
                            icon: "doc.text"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "api",
                            title: "API Reference",
                            icon: "link"
                        ),
                        TchatBreadcrumb.BreadcrumbItem(
                            id: "endpoints",
                            title: "Endpoints",
                            isClickable: false
                        )
                    ],
                    style: .standard,
                    size: .large,
                    separator: "arrow.right",
                    showHome: true,
                    homeIcon: "house.fill"
                )
            }
            .padding()
        }
        .background(Color(.systemBackground))
        .previewLayout(.sizeThatFits)
    }
}
#endif